qiskit_nature/second_q/operators/tensor.py

```diff
@@ -0,0 +1,409 @@
+
+        If the instance on which this method was called already fulfilled this requirement, it is
+        returned unchanged.
+
+        Args:
+            sparse_type: the type to use for the conversion to sparse array. Note, that this will
+            only be applied if the wrapped array object was dense. Converting an already sparse
+            array to another sparse type needs to be done explicitly.
+
+        Returns:
+            A new ``Tensor`` with the internal array converted to the requested sparse array type.
+        """
+        if self.is_sparse():
+            return self
+
+        return self.__array_wrap__(sparse_type.from_numpy(self._array))
+
+    def to_dense(self) -> Tensor:
+        """Returns a new instance with the internal array converted to a dense numpy array.
+
+        If the instance on which this method was called already fulfilled this requirement, it is
+        returned unchanged.
+        """
+        if self.is_dense():
+            return self
+
+        return self.__array_wrap__(cast(SparseArray, self._array).todense())
+
+    # pylint: disable=too-many-return-statements
+    def __eq__(self, other: object) -> bool:
+        """Check equality of ``Tensor`` instances.
+
+        .. note::
+            This check only asserts the internal matrix elements for equivalence but ignores the
+            type of the matrices. As such, it will indicate equality of two ``Tensor`` instances
+            even if one contains sparse and the other dense numpy arrays, as long as their elements
+            are identical.
+
+        Args:
+            other: the second ``Tensor`` object to be compared with the first.
+
+        Returns:
+            True when the ``Tensor`` objects are equal, False when not.
+        """
+        if not isinstance(other, (Tensor, Number, np.ndarray, SparseArray)):
+            return False
+
+        value = Tensor(self)._array
+        other_value = Tensor(other)._array
+
+        self_is_sparse = isinstance(value, SparseArray)
+        other_is_sparse = isinstance(other_value, SparseArray)
+
+        if self_is_sparse:
+            if other_is_sparse:
+                if value.ndim != other_value.ndim:
+                    return False
+                if value.nnz != other_value.nnz:  # type: ignore[union-attr]
+                    return False
+                if value.size != other_value.size:
+                    return False
+                diff = value - other_value
+                if diff.nnz != 0:
+                    return False
+                return True
+            value = value.todense()  # type: ignore[union-attr]
+        elif other_is_sparse:
+            other_value = cast(SparseArray, other_value).todense()
+
+        if not np.array_equal(value, other_value):
+            return False
+
+        return True
+
+    def __ne__(self, other: object) -> bool:
+        return not self.__eq__(other)
```

Comment:
Yes, this is necessary because the numpy mixin specifically sets this method differently: https://github.com/numpy/numpy/blob/3624b0628b9f493b7d2339a15ba3004dc90fc943/numpy/lib/mixins.py#L144